hierarchicalforecast/utils.py

```diff
@@ -77,384 +80,459 @@ def cov2corr(cov, return_std=False):
-
-    S = encoder.fit_transform(S_df).T
-
-    if sparse_s:
-        df_constructor = pd.DataFrame.sparse.from_spmatrix
-    else:
-        df_constructor = pd.DataFrame
-    S = df_constructor(S, index=chain(*categories), columns=cats_bottom)
-
-    tags = dict(zip(S_df.columns, categories))
-    return S, tags
-
 # %% ../nbs/src/utils.ipynb 11
-def aggregate_before(df: pd.DataFrame,
-              spec: List[List[str]],
-              agg_fn: Callable = np.sum,
-              sparse_s: bool = False):
-    """Utils Aggregation Function.
-
-    Aggregates bottom level series contained in the pd.DataFrame `df` according
-    to levels defined in the `spec` list applying the `agg_fn` (sum, mean).<br>
-
-    **Parameters:**<br>
-    `df`: pd.DataFrame with columns `['ds', 'y']` and columns to aggregate.<br>
-    `spec`: List of levels. Each element of the list contains a list of columns of `df` to aggregate.<br>
-    `agg_fn`: Function used to aggregate `'y'`.<br>
-    `sparse_s`: bool=False, whether the returned S should be a sparse DataFrame.<br>
-
-
-    **Returns:**<br>
-    `Y_df, S, tags`: tuple with hierarchically structured series `Y_df` ($\mathbf{y}_{[a,b]}$),
-    summing matrix `S`, and hierarchical aggregation indexes `tags`.
-    """
-    max_len_idx = np.argmax([len(hier) for hier in spec])
-    bottom_comb = spec[max_len_idx]
-    hiers = []
-    for hier in spec:
-        df_hier = df.groupby(hier + ['ds'])['y'].apply(agg_fn).reset_index()
-        df_hier['unique_id'] = df_hier[hier].agg('/'.join, axis=1)
-        if hier == bottom_comb:
-            bottom_hier = df_hier['unique_id'].unique()
-        hiers.append(df_hier)
-    df_hiers = pd.concat(hiers)
-    S_df = df_hiers[['unique_id'] + bottom_comb].drop_duplicates().reset_index(drop=True)
-    S_df = S_df.set_index('unique_id')
-    S_df = S_df.fillna('agg')
-    hiers_cols = []
-    for hier in spec:
-        hier_col = '/'.join(hier)
-        S_df[hier_col] = S_df[hier].agg('/'.join, axis=1)
-        hiers_cols.append(hier_col)
-    Y_df = df_hiers[['unique_id', 'ds', 'y']].set_index('unique_id')
-    
-    # Aggregations constraints S definition
-    S, tags = _to_summing_matrix(S_df.loc[bottom_hier, hiers_cols], sparse_s)
-    return Y_df, S, tags
-
-# %% ../nbs/src/utils.ipynb 12
 def _to_upper_hierarchy(bottom_split, bottom_values, upper_key):
-    upper_split = upper_key.split('/')
+    upper_split = upper_key.split("/")
     upper_idxs = [bottom_split.index(i) for i in upper_split]
 
     def join_upper(bottom_value):
-        bottom_parts = bottom_value.split('/')
-        return '/'.join(bottom_parts[i] for i in upper_idxs)
+        bottom_parts = bottom_value.split("/")
+        return "/".join(bottom_parts[i] for i in upper_idxs)
 
     return [join_upper(val) for val in bottom_values]
 
-# %% ../nbs/src/utils.ipynb 13
+# %% ../nbs/src/utils.ipynb 14
 def aggregate(
-    df: pd.DataFrame,
+    df: Frame,
```

Comment:
you could use `FrameT` here to annotate the return as `tuple[FrameT, FrameT, dict]`